### DISCUSSION TITLE
Casting an attribute

### DISCUSSION THREAD

**Discussion by amacfie-tc** [microsoft/pyright#9581] [Q&A] [answered]
```python
from typing import Any, cast, reveal_type


class C:
    a: dict[int, int] | None = None


c = C()
c.a = cast(dict[int, Any], c.a)
reveal_type(c.a)  # Type of "c.a" is "dict[int, int]"
```
What can a `cast` call be assigned to for Pyright to understand? Just atomic names like `foo`?

**Comment by erictraut** [comment, marked as answer]
A `cast` doesn't affect assignment. The subexpression `cast(dict[int, Any], c.a)` simply tells a type checker to ignore the type it normally evaluates for the `c.a` and replace that type with `dict[int, Any]`. 

When you assign a value of type `dict[int, Any]` (whether that comes from normal type evaluation or a `cast`) to an attribute whose type is declared as `dict[int, int] | None`, the assignment is allowed because `dict[int, Any]` is assignable to `dict[int, int] | None`. [Type narrowing](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing) is applied on assignment. The type `dict[int, int] | None` is narrowed to `dict[int, int]` in this case. The `reveal_type` call displays the locally-narrowed type at that location within the program.
